text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+We will go into more detail about these steps below.
+
+### Social consensus
+
+For the initial version of this CEP, we assume that we have some kind of platform that enables open communication between validators and CasperLabs. The exact details of what that platform will be (e-mail? Discord? a forum?) can be left to be determined in the future.
+
+The first essential step will be determining the block to be used as the starting point for the network after the restart. There are a few main options here:
+
+1. It could be the last block whose FTT still exceeds the total weight of the equivocating validators - in which case it could be done automatically, as mentioned above.
+2. Another option would be to use the tip of one of the forks of the chain, if more than one fork has been created before the detection of the equivocation catastrophe.
+3. The last option would be to use a manually constructed block (more details below).
+
+A desirable goal here would be to revert as few transactions from all the forks as possible, in order to limit the damage. Every transaction that was included in a block which appeared to be finalized for a time could have caused real world consequences (like, for example, a merchant sending ordered goods). Reverting any such transaction might potentially cause a loss to a user.
+
+Option 1 essentially reverts all transactions that happened after the block whose FTT has been exceeded. Option 2 attempts to revert the transactions from all forks but one. However, it is possible that some of the transactions on different forks of the chain don't conflict with each other. In such a case, we could attempt to replay all of them on the new trusted branch. This is the idea behind option 3: to gather the non-conflicting transactions, create a block containing them and collectively declare it correct, outside of the consensus algorithm.
+
+After the users collectively decide upon the last correct block, the transactions to be replayed and the list of equivocators to be slashed, the network can be restarted.
+
+### Restarting the network
+
+The process of restarting the network could be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators, which is based on the bonded validators of the last correct block, with the equivocators removed.
+- There is a different initial global state: it is not empty anymore, but it is the post-state of the last correct block.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
```

Comment:
Catastrophes are generally events that require validators to react very quickly, in a matter of hours. I think the reference-level implementation needs to include the time windows for different steps once the catastrophe gets acknowledged. E.g. maximum duration validators will wait to tally votes for step 3.

This way, there would be no liability of participating validators, if certain validators are too late to react, and try to object to the decision because they don't like to outcome.

I don't remember how fast the hard fork for the DAO hack was implemented, but if [this post](https://medium.com/@slacknation/dao-hack-timeline-823e5a18e894) is to be believed, Vitalik announced the recovery plan a bit under 8 hours following the attack.

Reply:
I think this CEP could use a few more details about comms procedures, possible uses of Clarity UI, etc., otherwise we would just need to do all of that in another CEP. Alternatively, we could make this about selection of restart point/GS _only_ and do the restart itself in a later CEP? @fizyk20

Reply:
Especially relevant since there appears to be a dependency on CEP-5, which isn't done yet.

Reply:
Essentially, I want to keep this one as vague as possible, since the details could depend on the particular scenario and when it happens (which could influence what tools we have available, what is the main channel of communication with the community and various other things). The idea is to outline a general process, with the details being filled in when more is known about the specific case.

Reply:
Yes, I would imagine it would be a new chainspec.toml + any dependencies, so accounts.csv too. I believe that we should provide the necessary tools, yes (it is already mentioned in the latest version of the CEP).